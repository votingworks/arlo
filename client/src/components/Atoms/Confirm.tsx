import React, { useState, ReactNode } from 'react'
import { Dialog, Classes, Button, Intent } from '@blueprintjs/core'

export interface IConfirmOptions {
  title: ReactNode
  description: ReactNode
  yesButtonLabel?: string
  yesButtonIntent?: Intent
  noButtonLabel?: string
  onYesClick: () => Promise<void>
}

export const useConfirm = () => {
  // We show the dialog whenever options are set.
  // On close, we set options to null.
  const [options, setOptions] = useState<IConfirmOptions | null>(null)

  const confirm = (newOptions: IConfirmOptions | null) => {
    setOptions(newOptions)
  }

  const onYesClick = async () => {
    await options!.onYesClick()
    setOptions(null)
  }

  const onClose = () => {
    setOptions(null)
  }

  const confirmProps = {
    isOpen: !!options,
    title: options ? options.title : '',
    description: options ? options.description : '',
    yesButtonLabel: options ? options.yesButtonLabel : undefined,
    yesButtonIntent: options ? options.yesButtonIntent : undefined,
    noButtonLabel: options ? options.noButtonLabel : undefined,
    onYesClick,
    onClose,
  }

  return { confirm, confirmProps }
}

interface IConfirmProps extends IConfirmOptions {
  isOpen: boolean
  onClose: () => void
}

export const Confirm = ({
  isOpen,
  title,
  description,
  yesButtonLabel,
  yesButtonIntent,
  noButtonLabel,
  onYesClick,
  onClose,
}: IConfirmProps) => {
  const [isSubmitting, setIsSubmitting] = useState<boolean>(false)

  const handleYesClick = async () => {
    setIsSubmitting(true)
    try {
      await onYesClick()
    } catch (error) {
      // Do nothing, error handling should happen within onYesClick
    } finally {
      setIsSubmitting(false)
    }
  }

  return (
    <Dialog icon="info-sign" onClose={onClose} title={title} isOpen={isOpen}>
      <div className={Classes.DIALOG_BODY}>{description}</div>
      <div className={Classes.DIALOG_FOOTER}>
        <div className={Classes.DIALOG_FOOTER_ACTIONS}>
          <Button disabled={isSubmitting} onClick={onClose}>
            {noButtonLabel || 'Cancel'}
          </Button>
          <Button
            intent={yesButtonIntent || Intent.PRIMARY}
            onClick={handleYesClick}
            loading={isSubmitting}
          >
            {yesButtonLabel || 'Ok'}
          </Button>
        </div>
      </div>
    </Dialog>
  )
}
