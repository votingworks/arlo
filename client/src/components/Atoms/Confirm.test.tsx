import React from 'react'
import { render, screen, within, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { Button } from '@blueprintjs/core'
import { useConfirm, Confirm } from './Confirm'

const onYesClickMock = jest.fn().mockResolvedValue(true)

const ConfirmConsumer = () => {
  const { confirm, confirmProps } = useConfirm()
  return (
    <>
      <Button
        onClick={() =>
          confirm({
            title: 'Test Title',
            description:
              'This action cannot be undone. Do you want to continue?',
            onYesClick: onYesClickMock,
            yesButtonLabel: 'Yes',
          })
        }
      >
        Open Confirm
      </Button>
      <Confirm {...confirmProps} />
    </>
  )
}

afterEach(() => {
  jest.clearAllMocks()
})

describe('Confirm component', () => {
  it('opens confirm component', async () => {
    render(<ConfirmConsumer />)

    userEvent.click(screen.getByRole('button', { name: 'Open Confirm' }))

    const dialog = (await screen.findByRole('heading', {
      name: /Test Title/,
    })).closest('.bp3-dialog')! as HTMLElement
    within(dialog).getByText(
      'This action cannot be undone. Do you want to continue?'
    )
  })

  it('closes confirm component when cancelled', async () => {
    render(<ConfirmConsumer />)

    userEvent.click(screen.getByRole('button', { name: 'Open Confirm' }))

    const dialog = (await screen.findByRole('heading', {
      name: /Test Title/,
    })).closest('.bp3-dialog')! as HTMLElement
    within(dialog).getByText(
      'This action cannot be undone. Do you want to continue?'
    )

    userEvent.click(within(dialog).getByRole('button', { name: 'Cancel' }))

    await waitFor(() => {
      expect(
        screen.queryByText(
          'This action cannot be undone. Do you want to continue?'
        )
      ).not.toBeInTheDocument()
    })
  })

  it('closes confirm component when clicked on Yes', async () => {
    render(<ConfirmConsumer />)

    userEvent.click(screen.getByRole('button', { name: 'Open Confirm' }))

    const dialog = (await screen.findByRole('heading', {
      name: /Test Title/,
    })).closest('.bp3-dialog')! as HTMLElement
    within(dialog).getByText(
      'This action cannot be undone. Do you want to continue?'
    )

    userEvent.click(within(dialog).getByRole('button', { name: 'Yes' }))

    expect(onYesClickMock).toHaveBeenCalled()

    await waitFor(() => {
      expect(
        screen.queryByText(
          'This action cannot be undone. Do you want to continue?'
        )
      ).not.toBeInTheDocument()
    })
  })
})
