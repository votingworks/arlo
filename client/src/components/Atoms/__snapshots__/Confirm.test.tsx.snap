// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Confirm component closes confirm component 1`] = `<div />`;

exports[`Confirm component opens confirm component 1`] = `<div />`;
